### ISSUE TITLE
SC2034 false warning?

### ISSUE THREAD

**Issue by Skoti** [koalaman/shellcheck#995]
#### For bugs
- Rule Id: SC2034 
- My shellcheck version: 0.4.6
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet that shows the problem:

```sh

#!/bin/bash
array=("el1" "el2")
doSomething array[@]

```

#### Here's what shellcheck currently says:

^-- SC2034: array appears unused. Verify it or export it.

#### Here's what I wanted or expected to see:

I would assume this is a correct usage of a variable and see no warning.

**Comment by koalaman** [comment]
ShellCheck is right. `array` is not used here.

You can do `doSomething "${array[@]}"` to actually reference the array, otherwise it'll pass the string `array[@]`  literally (or glob expanded).

**Comment by Skoti** [comment]
@koalaman 
Thanks for an explanation, however, I'm passing an array as an argument into doSomething function and then I'm expanding it via indirection as:
```sh
declare -a array
array=("${!1}")
```
if I use `doSomething "${array[@]}"` then the code above no longer works, although I could get an array by `array=("$@")` but this means that if there are more arguments passed to doSomething function they are going to be included in the array, which is not desired.
Is there a better way of doing it?

I'm sorry if this is a newbie question/problem but I'm just starting with bash.

**Comment by koalaman** [comment]
It's shellcheck's intention not to try to figure out whether or not a variable is accessed through indirection.

The canonical way of doing this is to pass arrays as multiple positional parameters after optional flags, e.g. `myfunction -a foo -b bar "${args[@]}"`

If you still want to do it through indirection, ignore this warning.

**Comment by Skoti** [comment]
@koalaman 
ok got it ;), thanks for help!
